plumed: new port

GiovanniBussi opened the pull request:
Dear MacPorts developers,

I am opening a pull request for a new port submission related to the software [PLUMED](http://www.plumed.org), which I am co-developing. Before asking you to merge it, I would like some feedback about the way I structured this rather complex Portfile so that I could perhaps simplify it a bit.

**So, please do not merge yet!**

I list my questions here:

- Even though the PLUMED software resides on [github](http://github.com/plumed/plumed2), I preferred to use manually uploaded .tgz files. This allowed me to save checksums (thus avoiding new downloads when files have been fetched already) and to split the download in two separate .tgz files (plumed-src and plumed-doc). The latter is very big and is only downloaded when installing documentation (variant +doc). Is this ok?
- I defined several variants so as to switch on/off dependent libraries (+xdrfile, +matheval, +zlib, +gsl). Some of them (+xdrfile, +matheval, and +zlib) are active by default since they require small libraries that are fast to install. Should I remove these three variants completely and make the dependence hard coded? I would prefer to keep gsl optional since it needs time to compile.
- PLUMED itself contains modules that might have different authors. I am adding variants for each of the non-default modules (here +mod_manyrestraints, +mod_crystallization, +mod_adjmat). The number of modules will increase in the future. Is it bad practice to have a variant for each of them? Should I only keep the variant +allmodules which enables all of them?
- I have a variant +molfile which can be used to discard (with -molfile) an internal piece of code with a BSD license. I thought someone might want to have a pure LGPL3 license. Is there any advantage or I can also remove this variant to simplify the Portfile?
- I did not manage to setup tests. I could do it if necessary, since we have a wide spectrum regression test suite contained in an additional plumed-test tgz file. Is it possible to selectively download it when the user wants to run the test?

Finally, notice that this port is based on just released PLUMED 2.3.1 and requires a small patch to work smoothly with MacPorts (already included in files/). Patch is already included upstream and will be removed when we will release PLUMED 2.3.2.

Many thanks in advance for your help and sorry for the long post!

Giovanni

###### Description
Implement a Portfile for PLUMED software.

###### Tested on
macOS 10.11.6
Xcode 8.1

###### Verification <!-- (delete not applicable items) -->
Have you
- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [X] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?

mojca commented:
Hi. Thanks a lot for your contribution. I wanted to warn you about one specific thing. Assuming you'll manage to make the port binary distributable, changing any variant will require building the port from source. So if you can afford to make multiple subports that can be installed independently, that might be preferred.

> - Even though the PLUMED software resides on [github](http://github.com/plumed/plumed2), I preferred to use manually uploaded .tgz files. This allowed me to save checksums (thus avoiding new downloads when files have been fetched already) and to split the download in two separate .tgz files (plumed-src and plumed-doc). The latter is very big and is only downloaded when installing documentation (variant +doc). Is this ok?

Yes. But as I have written above: if users will want the documentation installed and the default will be without the docs, users will have to build the port from source rather than being able to fetch the binaries directly. Unless the documentation is hardwired and one cannot build the documentation separately, it might be wise to split this into separate ports.

> - I defined several variants so as to switch on/off dependent libraries (+xdrfile, +matheval, +zlib, +gsl). Some of them (+xdrfile, +matheval, and +zlib) are active by default since they require small libraries that are fast to install. Should I remove these three variants completely and make the dependence hard coded? I would prefer to keep gsl optional since it needs time to compile.

Keep it simple. Changing a single variant requires rebuilding from source. Unless the user really needs to have `zlib` disabled for whatever obscure reason, add it in by default and forget about a separate variant.

Also keep in mind that packages hopefully come in binary form. [http://packages.macports.org/gsl/ gsl] seems to be just 2 MB, so I don't see a problem with including it by default if it adds additional functionality.

> - PLUMED itself contains modules that might have different authors. I am adding variants for each of the non-default modules (here +mod_manyrestraints, +mod_crystallization, +mod_adjmat). The number of modules will increase in the future. Is it bad practice to have a variant for each of them? Should I only keep the variant +allmodules which enables all of them?

Keep my first paragraph in mind. Can you install individual modules independently? If so, you could create a subport `plumed-manyrestrains` etc. Then users will be free to install and uninstall modules without having to recompile the main port.

The final decision depends on you of course. We don't know how the software is used and what your users need, but having too many variants would usually strike you back.

> - I have a variant +molfile which can be used to discard (with -molfile) an internal piece of code with a BSD license. I thought someone might want to have a pure LGPL3 license. Is there any advantage or I can also remove this variant to simplify the Portfile?

I don't know if BSD is compatible with LGPL3. I thought it was less restrictive, but I'm not eager to investigate now. You need to specify the licence as one of the fields and I'm not sure what's the proper value in case the software has multiple licences at the same time (it depends a bit).

There is a script `/opt/local/var/macports/sources/rsync.macports.org/release/tarballs/base/portmgr/jobs/port_binary_distributable.tcl` that lets you know whether your port is binary distributable or not (call it with `-v` for verbose). You should try to figure out if you could make the port binary distributable.

> - I did not manage to setup tests. I could do it if necessary, since we have a wide spectrum regression test suite contained in an additional plumed-test tgz file. Is it possible to selectively download it when the user wants to run the test?

Can you please ask on the macports-dev mailing list?

mojca commented on science/plumed/Portfile:
```diff
@@ -0,0 +1,140 @@
+PortGroup           debug 1.0
+
+version             2.3.1
+revision            0
```
When version is 0, leave this out.

mojca commented on science/plumed/Portfile:
```diff
@@ -0,0 +1,140 @@
+
+homepage            http://www.plumed.org/
+
+master_sites        https://github.com/plumed/plumed2/releases/download/v${version}/
```
I think there's a shortcut for this, but I should take a look.

mojca commented on science/plumed/Portfile:
```diff
@@ -0,0 +1,140 @@
+          rmd160 28bbe50a3a188419ce3ab1c09aff4f95bd0595c9 \
+
+# This is required so that patch made with "git format-patch" work correctly
+patch.pre_args -p1
```
This is maintainer's choice. I usually delete `a/` and `b/` manually to avoid having to do this.

mojca commented on science/plumed/Portfile:
```diff
@@ -0,0 +1,140 @@
+default_variants +matheval +xdrfile +zlib +molfile
+
+# This is required since PLUMED installation does not do it explicitly.
+# It might be removed when this will be incorporated upstream
```
Yes, fix this upstream please. You may temporary do this, but this should really be fixed. (Don't autotools automatically take care of this?)

GiovanniBussi commented:
Thanks @mojca for your feedbacks! I implemented most of them and made some further simplification. Also thanks for suggesting the script that checks for binary redistribution.

In particular:
- I removed +doc variant for several reasons:
  * This variant would not be binary redistributable (it needs doxygen which is GPL2 and not compatible with PLUMED).
  * The same manual is already available online.
  * Having an ad hoc manual would only be useful for advanced users who are compiling plumed by themselves.
- I removed the variants for internal modules. I will think at the subport solution, that looks very neat, thanks for suggesting. So far I will only keep the extra modules optionals with a single variant +allmodules
- I removed all the variants associated to external libraries. You are right, also GSL is very small. In case one installs plumed with +gcc6, GSL should be compiled from source. Also in this case, it only takes a few minutes.
- I removed the -molfile variant and included the BSD code by default.
- About tests, I though not to include them. Our tests can be also performed on an installed version, so it is probably better to just let the user perform them outside of MacPorts.
- About the install_name_tool: we are not using libtool, but I will anyway fix this upstream.

Many thanks again for your help. If you think the Portfile is ok, feel free to merge it.

Giovanni

GiovanniBussi commented:
Dear MacPorts developers,

would it be possible to merge this new portfile?

In case it still has some problem I am happy to improve it.

Many thanks!

Giovanni

GiovanniBussi commented:
Dear @mojca,

this new port is still pending. Would it be possible to merge it? If not, is there something I can do to improve it?

Thanks!

Giovanni

mojca commented on science/plumed/Portfile:
```diff
@@ -0,0 +1,88 @@
+post-destroot {
+    system "install_name_tool -id ${prefix}/lib/libplumed.dylib ${destroot}${prefix}/lib/libplumed.dylib"
+    system "install_name_tool -id ${prefix}/lib/libplumedKernel.dylib ${destroot}${prefix}/lib/libplumedKernel.dylib"
+}
```
Autotools should take care of doing this automatically if implemented properly.

mojca commented on science/plumed/Portfile:
```diff
@@ -0,0 +1,88 @@
+}
+
+# libraries
+configure.ldflags-append -lmatheval -lxdrfile -lz -lgsl
```
The build system should add those flags automatically.
Read as: the build system should be fixed upstream (it is acceptable to use such workarounds until this gets fixed, but it should really really be fixed eventually).

mojca commented on science/plumed/Portfile:
```diff
@@ -0,0 +1,88 @@
+# molfile plugins from VMD that are released with a BSD-like license
+# http://www.ks.uiuc.edu/Research/vmd/plugins/molfile/
+license             LGPL-3 BSD
+maintainers         gmail.com:giovanni.bussi openmaintainer
```
This could/should be
```
maintainers         {gmail.com:giovanni.bussi @GiovanniBussi} openmaintainer
```

mojca commented on science/plumed/files/0001-Added-enable-libsearch-to-configure.patch:
```diff
@@ -0,0 +1,420 @@
+   --enable-FEATURE[=ARG]  include FEATURE [ARG=yes]
+   --enable-modules        all/none/reset or : separated list such as
+                           +crystallization:-bias default: reset
++  --enable-libsearch      enable search for libraries, default: yes
```
This is a strange approach. Take a look at
https://autotools.io/pkgconfig/pkg_check_modules.html
or some other macros or check how other software deals with searching packages.

mojca commented on science/plumed/files/0001-Added-enable-libsearch-to-configure.patch:
```diff
@@ -0,0 +1,420 @@
+     found=
+-    for testlib in "" -lstdc++ -lc++ -lmpi_cxx ; do
++    if test "${libsearch}" == true ; then
++      testlibs="-lstdc++ -lc++ -lmpi_cxx"
```
This set of libraries is very weird (but I admit that I'm a bit ignorant in this area). Are you simultaneously including `libc++` and `libstdc++`?

If that is the case, you should respect the variable that's given to you with `${configure.cxx_stdlib}`.

mojca commented:
Oh, you are presenting the software in Trieste next week?
In that case just drop by https://www.mittelab.org/ @g5pw to get any issues fixed faster :) :) :)

GiovanniBussi commented:
@mojca sorry for the delay. Some of the issues you raised are solved upstream and will be fixed in PLUMED 2.3.2, to be released in a few weeks. I prefer to wait for that instead of submitting a new Portfile with too many patches. Meanwhile let me answer to some of your points:

- We do not use autotools, but rather autoconf alone (with a number of _ad hoc_ tests) in combination with GNU make. We do not use libtools either. Additionally, we do  not rely on pkg-config since we are not sure how portable the approach would be (in some HPC machine I used it was not available). We __might__ start using libtools and perhaps other autotools and pkg-config in the future, but certainly not early than PLUMED 2.4 (released end of 2017), and perhaps even later. So, in case using autotools and pkg-config is a requirement to be merged in macports trunk I     would rather close the pull request and reopen it later when (if) these concerns are resolved.

- `configure.ldflags-append -lmatheval -lxdrfile -lz -lgsl`. I know this is odd, but I feel that this is much more robust than the usual policy of automatically adding standard library names. This is something I added specifically for PLUMED to be included in MacPorts since I though it is    more robust, but I can remove it easily if you think it is a bad idea (see below). The problem I want to avoid is the following. Imagine someone adds a new optional library to be     searched in PLUMED configure (say `libhello.dylib`), and that this library is also available in MacPorts as package `libhello` but we forget to explicitly add   it as `depends_lib-append port:libhello`. Then, when PLUMED happens to be installed from source by a user that by chance has `libhello.dylib` installed,         `libplumed.dylib` will be linked to that. If the user later removes the `libhello` package, the `libplumed.dylib` will not work anymore. To avoid this, I added  a special flag (`--disable-libsearch`) that tells `./configure` not to add library names to library flags. Only those that are explicitly listed in `LDFLAGS`    will be linked. The reason why I care about this is that in at least two cases I found issues related to the choice of automatic adding `-l` flags within macports (see #409, where it was necessary to explicitly instruct configure to use an internal package instead of the external one, and #394, where a dependence was missing). If you don't like this for some reason, I can change the behavior by simply:
  -  removing `--disable-libsearch` and
  -  removing `configure.ldflags-append -lmatheval -lxdrfile -lz -lgsl`
  
Obviously, in this case the names will be automatically added by configure as you expected (and as we do on other platforms). What do you think?

In any case, in a few days/weeks I will submit a new Portfile based on PLUMED 2.3.2 where I will try to explain in detail all these choices in the comments inside the Portfile, so that the motivation is clear.

@g5pw I actually live in Trieste (and work at SISSA).

Thanks again!

Giovanni

raimue commented:
> Then, when PLUMED happens to be installed from source by a user that by chance has libhello.dylib installed, libplumed.dylib will be linked to that. If the user later removes the libhello package, the libplumed.dylib will not work anymore.

Missing dependencies of this kind are usually caught on the buildbots, because they only activate the dependening ports. Furthermore, you can test this yourself with trace mode (`sudo port -t build ...`), which prevents access to files not provided by dependencies.

GiovanniBussi commented:
@raimue thanks for the comment. I indeed noticed this the first time I used `port -t build`. Though, the fact that in at least two ports (`grace`, #409, and `libmatheval`, #394) the problem was not detected for a long time makes me think that the procedure I am using would be more robust.

I will keep it in the Portfile I will resubmit, though I am happy to change it (removing `--disable-libsearch` and `configure.ldflags-append -lmatheval -lxdrfile -lz -lgsl`) if you think it is better. In this case, I will find a solution upstream to make sure that extra libraries are disabled by hand (adding `--disable-libhello` in the plumed Portfile whenever someone adds a new library `libhello` to be searched in our `configure.ac`).

GiovanniBussi commented:
Dear @mojca and @raimue,

We just released PLUMED 2.3.2 upstream and the Portfile should be ready for merging. The Portfile is extensively commented for future reference. I think I already answered to all your questions in the thread above, but let me summarize here the most important points:
1. We are not using autotools, but rather autoconf with a [custom configure.ac](https://github.com/plumed/plumed2/blob/v2.3.2/configure.ac). This is why you see some non-standard configure options.
2. I am still using a custom configure flag `--disable-libsearch` to avoid searching for libraries. In this manner, only libraries explicitly added to `configure.ldflags-append` will be linked. I think this is more robust as it allows the Portfile to be totally aware of any linked library. __If necessary I can change this__ by removing `--disable-libsearch` and `configure.ldflags-append -lmatheval -lxdrfile -lz -lgsl`
3. I am using several standard portgroups, and I guess it is not necessary to document the related variants (e.g. `+gcc6` or `+openmpi`). The only custom variant is `+allmodules` that turns on some optional part of PLUMED. These parts are either contributed by other developers or, for some reason, considered as less stable, so we prefer them not to be included by default. Installation from source with `port install plumed +allmodules` is anyway not too long (a few minutes).

Could you please go ahead and merge or let me know if there is something else I should do?

Many thanks!

Giovanni

GiovanniBussi commented:
Dear all,

would it be possible to merge this new portfile?

In there is some problem, I am happy to make further changes.

Many thanks!

Giovanni

mf2k commented:
I think you should move the docs into a sub-port (not a variant) which allows you to use the Github portgroup and reduce the download size. That is the best of both worlds.

GiovanniBussi commented:
Dear @mf2k, thanks for your comment.

@mojca already gave me this hint, and I thus decided to remove the `+doc` variant. I also thought about the possibility to have a `plumed-doc` subport, but I think it is not worth doing it since the manual is very large and is [available online](https://plumed.github.io/doc-v2.3). So, in the current Portfile, documentation is not installed.

Concerning the Github portgroup, the reason why I did not use it is that in my experience it leads to a new download for every new install. That is, the system assumes that the image could have been changed and downloads it again. In addition, the file that I have to download is not the whole zip archive but a smaller archive uploaded in the download tab of Github (e.g. file `plumed-src-2.3.2.tgz` [here](https://github.com/plumed/plumed2/releases/tag/v2.3.2)). I didn't manage to instruct macports to download this file using the Github portgroup. Notice that this file is way smaller than the whole archive (~3MB vs ~ 35 MB) since it does not contains tests nor the manual, so I would like to stick to this smaller archive.

Is there any specific advantage in using the Github portgroup?

Thanks!

Giovanni

mf2k commented:
I tested it and I don't experience a "new download for every new install" when using the GitHub portgroup. Why would that be an issue anyway? In general, once a port is installed, you would only re-install on update. At that point, you must download the new archive. 

Using the GitHub portgroup simplifies the Portfile. Going forward, changes in GitHub will be handled by the portgroup and thus all ports gain the advantage of that without having to be fixed individually.

I just don't see saving 30 MB of space as a big deal. You can clear your local distfiles if you want. But modifying the GitHub Portgroup to download a "src" archive as an option seems like the best solution so that other ports can potentially gain this advantage. Yet another reason to use the GitHub portgroup. :)

GiovanniBussi commented:
Dear @mf2k ,
  
thanks for your advise.

If I understand correctly there are three options:

1. Leave things as they are now, that is downloading `plumed-src-version.tar.gz` without using the github portgroup.
2. Improve the github portgroup to allow downloading tar.gz files with arbitrary names.
3. Go back to full download of the default archive created by github (thus downloading the whole archive), thus using github portgroup.

If I understand correctly, you consider no. 1 as bad practice. My knowledge of tcl is too primitive to implement no. 2. I will thus go for  no. 3, that is using the simpler github portgroup options. You can find an updated version on this same branch now.

Concerning my previous comment about fetching the file twice, I double checked and you are right, the tar.gz is not fetched again. I am not sure why this was happening to me some time ago, maybe I was doing another mistake.

Let me know if there is something else I should fix.

Thanks!

Giovanni

PS I see that now pull requests are tested on Travis-CI. This is really useful, great job!

mf2k commented:
Yes, we consider option 1 bad practice. 

For 2, we don't want to allow arbitrary filenames, but instead add an option to download the src filename. So basically it would be a line something like this:

github.fetch_src_file  yes

And the GitHub portgroup would do its magic to fetch plumed-src-2.4a.tgz instead of plumed-2.4a.tgz. 

I'm not clear how to make the change myself either. So I think option 3 is best for now.

GiovanniBussi commented:
OK great.

So if my Portfile is ok could you merge it? Should I rebase it as a single commit?

Thanks!

Giovanni

mf2k commented:
Sorry for the delay on this. Please remove the extraneous comments about patch files and rebase.

GiovanniBussi commented:
@mf2k no problema at all, thanks for taking care of this.

I removed the comments, rebased on master branch and squashed. Hopefully this is ok.

Thanks again

Giovanni

GiovanniBussi commented:
Dear @mf2k, thanks for merging. I just received an email saying that plumed does not build (see [here](https://build.macports.org/builders/ports-10.6_i386_legacy-watcher/builds/8011) and [here](https://build.macports.org/builders/ports-10.6_i386_legacy-builder/builds/25416/steps/install-port/logs/stdio)). Should I do something? I guess the problem is with g++-4.2, which I never tested.

mf2k commented:
That's just the legacy builder for Snow Leopard. We don't officially support it, so if you have the ability to debug old operating systems and want to spend the time on it, then go for it.

GiovanniBussi commented:
@mf2k thanks! I will try later on to see if I can reproduce the bug with g++-4.2 and fix it. Otherwise I will leave it as this.

Thanks again to all MacPorts developer, you are doing a great job

Giovanni
